tutorials/arvo/cli-tutorial.md

```diff
@@ -0,0 +1,463 @@
+matches `demo`.
+
+The reason we use a `cord` `demo` as the argument for `jest` rather than a
+`tape` is not too important for this discussion, but the short version is that
+`cord`s are used to represent data that is passed around while `tape`s are for
+when you're ready to start doing string manipulation.
+
+```hoon
+++  tab-list
+  |=  sole-id=@ta
+  ^-  (list [@t tank])
+  :~  ['demo' leaf+"run example command"]
+  ==
+```
+`+tab-list` is pretty much plug-n-play. For each command you want to be tab
+completed, add an entry to the `list` begun by `:~` of the form `[%command
+leaf+"description"]`. Now whenever the user types a partial command and presses
+tab, the console will display the list of commmands that match the partial
+command as well as the descriptions given here.
+
+Thus here we have that starting to type `demo` and pressing tab will result in
+the following output in the console:
+```
+demo  run example command
+~zod:shoe> demo
+```
+with the remainder of `demo` now displayed in the input line.
+
+Next we have `+on-command`, which is called whenever `+command-parser`
+recognizes that `demo` has be entered by a user on a connected session.
+```hoon
+++  on-command
+  |=  [sole-id=@ta =command]
+  ^-  (quip card _this)
+```
+This is a gate that takes in the `sole-id` corresponding to the session and the
+`command` noun parsed by `+command-parser` and returns a `list` of `card`s and
+`_this`, which is Gall agent core including its state.
+```hoon
+  =-  [[%shoe ~ %sole -]~ this]
+```
+This creates a cell of a `%shoe` card that triggers a `sole-effect` given by the head of
+the subject `-`, then the Gall agent core `this` - i.e. the return result of
+this gate. The use of the `=-` rune means that what follows this
+expression is actually run first, which puts the desired `sole-effect` into the
+head of the subject.
+```hoon
+  =/  =tape  "{(scow %p src.bowl)} ran the command"
+```
+We pin the `tape` that we want to be printed to the head of the subject.
+```hoon
+  ?.  =(src our):bowl
+    [%txt tape]
+  [%klr [[`%br ~ `%g] [(crip tape)]~]~]
+```
+We cannot just put a bare `tape` at the head of the subject to be added to the
+card - it needs to be wrapped as a `sole-effect`. This tells us that if the
+origin of the command is not our ship to just print it normally with the `%txt`
+`sole-effect`. Otherwise we use `%klr`, which prints it stylistically (here it
+makes the text green and bold).
+
+The following allows either `~zod` or the current ship or its moons to connect to `%shoe`.
+```hoon
+++  can-connect
+  |=  sole-id=@ta
+  ^-  ?
+  ?|  =(~zod src.bowl)
+      (team:title [our src]:bowl)
+  ==
+```
+
+We use the minimal implementations for the final two `shoe` arms.
+```hoon
+++  on-connect      on-connect:des
+++  on-disconnect   on-disconnect:des
+--
+```
+
+This concludes our review of the code of the `%shoe` app. To continue learning
+how to build your own CLI app, we recommend checking out `/app/chat-cli.hoon`.
```

Comment:
Bad rec lol, chat-cli is old and complicated. Better cli apps Soon™.

Reply:
Is there something else I should suggest there or just leave it be for now?

Reply:
Probably leave this out for now? I think expecting users to extrapolate from /app/shoe isn't too big an ask, but maybe I'm underestimating the task. chat-cli is just going to be confusing and frustrating, there's way too much unrelated things in there.

Reply:
I've developed a lot of my own understanding over the past couple of weeks by browsing `chat-cli.hoon`. There is a lot of irrelevant stuff, but I think most users at this point should be able to pick it out. I don't think I would have gotten a better handle trying to extrapolate from `shoe.hoon`. Since there isn't anything else to point the reader at, for now I want to leave it in.